completion_aggregator/aggregator.py

```diff
@@ -0,0 +1,173 @@
+"""
+Aggregator service.
+
+This service periodically determines which stale_blocks need updating, and
+enqueues tasks to perform those updates.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import collections
+import logging
+import time
+from datetime import timedelta
+
+import six
+
+from django.db import transaction
+from django.utils import timezone
+
+from . import models, tasks, utils
+
+log = logging.getLogger(__name__)
+
+
+class BasePeriodicService(object):
+    """
+    Enqueue a new service every `period` for `runtime`, then quit.
+
+    `period` and `runtime` should be provided as timedelta objects, or None.  A
+    `period` of None indicates that the task should rerun immediately upon
+    completion.
+
+    Rather than starting this service from a cronjob every `period`, we provide
+    this service, which will iterate over to amortize the cost of firing up a
+    new edx-platform management task, which itself would take longer than a
+    reasonable period.
+    """
+
+    DEFAULT_PERIOD = timedelta(seconds=0)
+
+    service_name = "Unnamed Service"
+
+    def __init__(self, runtime=None, period=None):
+        """
+        Create a PeriodicService.
+
+        **Arguments:**
+
+        runtime (timedelta|None):
+            Run the service for this length of time, then quit.  If not
+            provided, run forever.
```

Comment:
I was thinking of it as a fallback to the service failing, but I suppose running it under supervisor would be a reasonable way to make sure it keeps running.

As for not running it under celerybeat, I was remembering some conversations with @jibsheet, where he said he didn't really like relying on celerybeat, and preferred using cron. I could be misremembering, though, as I see there is a job in CELERYBEAT_SCHEDULE:

```sh
(edx-platform) cliff@conakry:~/work/opencraft/docker/edx-platform$ rg CELERYBEAT lms/envs cms/envs
lms/envs/aws.py
102:CELERYBEAT_SCHEDULE = {}  # For scheduling tasks, entries can be added to this dict
699:        CELERYBEAT_SCHEDULE['refresh-saml-metadata'] = {
```

Reply:
If we do go to a celerybeat task, I still think it could be useful to have a management command that can be run manually when/if needed.  (Maybe with just the `run_once()` functionality.)

Reply:
Cool -- I'm happy to go with whatever edX would rather use.

When I looked at the celery docs, I thought we could use the code-driven [add_periodic_task](http://docs.celeryproject.org/en/latest/reference/celery.html#celery.Celery.add_periodic_task), but since we're on [celery 3.1](http://docs.celeryproject.org/en/3.1/reference/celery.html), we don't have that available.  So it looks like celerybeat, supervisord and cron would all require configuration changes.

Reply:
In the past our objection to celerybeat has been that it created a task per celery worker server, or required a standalone celerybeat server, neither of which is something we've wanted.  We tend to make jenkins jobs that run periodically and trigger a management command to queue a bunch of small tasks (see ACE for a recent example).
I see an email from Cliff sitting somewhere in my inbox, but since this was earlier I'll give a brief note on this.  I'm quite buried for this week, so may not get to the larger question before one of my teammates does.